File: frontend/src/core/algorithms.js
export class BFS {
  constructor(graph) {
    this.graph = graph;
    this.path = [];
    this.nodesExplored = 0;
    this.found = false;
  }

  run(startName, goalName) {
    this.graph.reset();
    this.path = [];
    this.nodesExplored = 0;
    this.found = false;

    const start = this.graph.getNode(startName);
    const goal = this.graph.getNode(goalName);
    if (!start || !goal) return [];

    const queue = [start];
    const visited = new Set([start.name]);
    start.parent = null;

    while (queue.length) {
      const current = queue.shift();
      this.nodesExplored += 1;

      if (current === goal) {
        this.found = true;
        this.path = this._reconstruct(goal);
        return this.path;
      }

      for (const { node: neighbor } of current.neighbors) {
        if (!visited.has(neighbor.name)) {
          visited.add(neighbor.name);
          neighbor.parent = current;
          queue.push(neighbor);
        }
      }
    }
    return [];
  }

  totalDistance() {
    if (this.path.length < 2) return 0;
    let total = 0;

    for (let i = 0; i < this.path.length - 1; i += 1) {
      const a = this.path[i];
      const b = this.path[i + 1];
      const nb = a.neighbors.find((n) => n.node.name === b.name);
      if (nb) total += nb.distance;
    }

    return total;
  }

  _reconstruct(goal) {
    const path = [];
    let current = goal;

    while (current) {
      path.push(current);
      current = current.parent;
    }

    return path.reverse();
  }
}

export class Dijkstra {
  constructor(graph) {
    this.graph = graph;
    this.path = [];
    this.nodesExplored = 0;
    this.found = false;
  }

  run(startName, goalName) {
    this.graph.reset();
    this.path = [];
    this.nodesExplored = 0;
    this.found = false;

    const start = this.graph.getNode(startName);
    const goal = this.graph.getNode(goalName);
    if (!start || !goal) return [];

    start.g = 0;
    const openList = [start];
    const closedSet = new Set();

    while (openList.length) {
      openList.sort((a, b) => a.g - b.g);
      const current = openList.shift();

      if (closedSet.has(current.name)) continue;
      closedSet.add(current.name);
      this.nodesExplored += 1;

      if (current === goal) {
        this.found = true;
        this.path = this._reconstruct(goal);
        return this.path;
      }

      for (const { node: nb, distance, traffic } of current.neighbors) {
        if (closedSet.has(nb.name)) continue;
        const tg = current.g + distance * traffic;
        if (tg < nb.g) {
          nb.g = tg;
          nb.parent = current;
          openList.push(nb);
        }
      }
    }

    return [];
  }

  _reconstruct(goal) {
    const path = [];
    let current = goal;

    while (current) {
      path.push(current);
      current = current.parent;
    }

    return path.reverse();
  }
}

File: core/algorithms.py
import heapq
from collections import deque
from core.heuristics import HEURISTICS


class BFS:
    def __init__(self, graph):
        self.graph          = graph
        self.path           = []
        self.nodes_explored = 0
        self.found          = False

    def run(self, start_name, goal_name):
        self.graph.reset()
        self.path           = []
        self.nodes_explored = 0
        self.found          = False

        start = self.graph.get_node(start_name)
        goal  = self.graph.get_node(goal_name)

        if not start or not goal:
            return []

        queue        = deque([start])
        visited      = set([start])
        start.parent = None

        while queue:
            current = queue.popleft()
            self.nodes_explored += 1

            if current == goal:
                self.found = True
                self.path  = self._reconstruct(goal)
                return self.path

            for neighbor, _, _ in current.neighbors:
                if neighbor not in visited:
                    visited.add(neighbor)
                    neighbor.parent = current
                    queue.append(neighbor)

        return []

    def _reconstruct(self, goal):
        path    = []
        current = goal
        while current:
            path.append(current)
            current = current.parent
        path.reverse()
        return path

    def total_distance(self):
        if not self.path or len(self.path) < 2:
            return 0
        total = 0
        for i in range(len(self.path) - 1):
            node_a = self.path[i]
            node_b = self.path[i + 1]
            for neighbor, dist, _ in node_a.neighbors:
                if neighbor == node_b:
                    total += dist
                    break
        return total


class Dijkstra:
    def __init__(self, graph):
        self.graph          = graph
        self.path           = []
        self.nodes_explored = 0
        self.found          = False

    def run(self, start_name, goal_name):
        self.graph.reset()
        self.path           = []
        self.nodes_explored = 0
        self.found          = False

        start = self.graph.get_node(start_name)
        goal  = self.graph.get_node(goal_name)

        if not start or not goal:
            return []

        start.g    = 0
        open_list  = [(0, start)]
        closed_set = set()

        while open_list:
            current_g, current = heapq.heappop(open_list)

            if current in closed_set:
                continue

            closed_set.add(current)
            self.nodes_explored += 1

            if current == goal:
                self.found = True
                self.path  = self._reconstruct(goal)
                return self.path

            for neighbor, distance, _ in current.neighbors:
                if neighbor in closed_set:
                    continue
                tentative_g = current.g + distance
                if tentative_g < neighbor.g:
                    neighbor.g      = tentative_g
                    neighbor.parent = current
                    heapq.heappush(open_list, (neighbor.g, neighbor))

        return []

    def _reconstruct(self, goal):
        path    = []
        current = goal
        while current:
            path.append(current)
            current = current.parent
        path.reverse()
        return path